### PR TITLE
[mlir][vector] Add tests xfer-permute-lowering (nfc)(2/n)

### PR DESCRIPTION
**Pull request by banach-space**
Adds more tests to:
  * vector-transfer-permutation-lowering.mlir

Specifically, adds tests for:
  * out-of-bounds access for the `TransferWritePermutationLowering`
    pattern
  * in-bounds access for `TransferWriteNonPermutationLowering` +
    `TransferWritePermutationLowering`

Also renames `@permutation_with_mask_xfer_write_fixed_width` as
`@xfer_write_non_transposing_permutation_map`.

This is a part of a larger effort to make sure that all key cases for
patterns under populateVectorTransferPermutationMapLoweringPatterns
(*) are tested. I also want to make sure that tests use consistent
function and variable names.

(*) transform.apply_patterns.vector.transfer_permutation_patterns in
TD parlance)

### CODE REVIEW

**Comment by banach-space** [comment]
@nujaa I've finally landed a few updates for `in_bounds` and can return to this - would you have cycles to take a look?

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 1]
```diff
@@ -83,19 +110,44 @@ func.func @xfer_write_transposing_permutation_map_masked(
...
+// CHECK-SAME:      %[[VEC:.*]]: vector<7xf32>,
+// CHECK-SAME:      %[[BASE_1:.*]]: index, %[[BASE_2:.*]]: index) {
+// CHECK:           %[[BC:.*]] = vector.broadcast %[[VEC]] : vector<7xf32> to vector<1x7xf32>
+// CHECK:           %[[TR:.*]] = vector.transpose %[[BC]], [1, 0] : vector<1x7xf32> to vector<7x1xf32>
+// CHECK:           vector.transfer_write %[[TR]], %[[MEM]]{{\[}}%[[BASE_1]], %[[BASE_2]]] {in_bounds = [false, true]} : vector<7x1xf32>, memref<?x?xf32>
+func.func @xfer_write_non_transposing_permutation_map(
+    %mem : memref<?x?xf32>,
+    %arg0 : vector<7xf32>,
+    %base1 : index,
+    %base2 : index) {
```
Comment (nujaa) [inline mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:124]: NIT: I'd suggest to use `%idx1` to match with other tests such as `masked_permutation_xfer_write_fixed_width` and your refacto of transfer-flatten.
I am not unhappy if you make it as part of a separate commit to tackle the above TODO. Also getting rid of  `%dim` args and use `%idx`.

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 2]
```diff
@@ -31,6 +33,31 @@ func.func @xfer_write_transposing_permutation_map(
   return
 }
 
+// Even with out-of-bounds, it is safe to apply this pattern
+// CHECK-LABEL:   func.func @xfer_write_transposing_permutation_map_out_of_bounds
+// CHECK-SAME:       %[[ARG_0:.*]]: vector<4x8xi16>,
```
Comment (nujaa) [inline mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:38]: NIT: Argument definitions are aligned with function body in the test below. Although `xfer_write_transposing_permutation_map` also follows this formatting. Both are valid I guess. 🤷

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 3]
```diff
@@ -1,5 +1,7 @@
 // RUN: mlir-opt %s --transform-interpreter --split-input-file | FileCheck %s
 
+// TODO: Replace %arg0 with %vec
```
Comment (nujaa) [inline mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:3]:
```suggestion
// TODO: Replace %arg0 with %vec
// TODO: Replace index args as %idx
// TODO: Align argument definition in CHECKS with function body.
```

**Review by nujaa — APPROVED** [review]
Hi, a couple of NITs depending on what you want to do with this list of MRs in the future. Thanks for pushing with this. LGTM. Suggested a change to keep track of potential other refactoring to take care in this file.

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir (hunk 3 above)
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:3]: Huge +1 - I will update the comment to reflect that.

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir (hunk 2 above)
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:38]: Let's prioritise consistency - I will update this before landing. Thanks!

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir (hunk 1 above)
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:124]: If you don't mind me expanding this PR, I'll do it here. Ultimately, this needs to happen and it's all about making the review process as smooth/easy as possible. If you are happy then I'm also happy :)

> Also getting rid of %dim args and use %idx.

This one I'm a bit unsure of - `%dim` is used for mask **dimension** rather than xfer_{read|write} index 🤔

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir (hunk 3 above)
[the remark this answers is not in the document]
Comment (nujaa) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:5]: I think you updated the file AND accepted my suggestion generating repetition 😕

[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:4]: Yikes! Fixed in the latest commit :)

**Comment by banach-space** [comment]
@nujaa Here's a follow-up to address some of your suggestions re formatting;
* https://github.com/llvm/llvm-project/pull/101468

More to come 😅
